### PR TITLE
Add cpu and memory stats to NodeTaskMap

### PR DESCRIPTION
**Pull request by cemcayiroglu**
NodeTaskMap holds the worker level aggregated split stats which is used by NodeSelectors. NodeSelector accesses to NodeTaskMap via NodeAssignmentStats. CPU and Memory stats are added to NodeTaskMap. CPU and Memory stats will used in planed new feature "workload placement".

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 1]
```diff
@@ -79,19 +103,33 @@ private NodeTasks addNodeTask(InternalNode node)
     private static class NodeTasks
     {
         private final Set<RemoteTask> remoteTasks = Sets.newConcurrentHashSet();
-        private final AtomicInteger nodeTotalPartitionedSplitCount = new AtomicInteger();
+        private final AtomicLong nodeTotalPartitionedSplitCount = new AtomicLong();
+        private final AtomicLong nodeTotalMemoryUsage = new AtomicLong();
+        private final AtomicLong nodeTotalCpuUsage = new AtomicLong();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:108]: Nit: add the units to the names (e.g., bytes, millis)

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 2]
```diff
@@ -108,87 +146,161 @@ private void addTask(RemoteTask task)
...
                 return toStringHelper(this)
                         .add("taskId", taskId)
-                        .add("splits", localPartitionedSplitCount)
+                        .add(stat, value)
                         .toString();
             }
         }
+
+        @ThreadSafe
+        private class TaskCumulativeStatsTracker
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:221]:
```suggestion
        private class CumulativeTaskStatsTracker
```

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 3]
```diff
@@ -108,87 +146,161 @@ private void addTask(RemoteTask task)
...
+                        Long firstReporting = values.firstKey();
+                        Long lastAge = values.lastKey();
+                        long delta = (value - values.get(firstReporting)) * 100;
+                        long duration = lastAge - firstReporting;
+                        if (delta >= 0 && duration > 0) {
+                            return delta / duration;
+                        }
+                        else {
+                            throw new IllegalArgumentException(stat + "delta is negative");
+                        }
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:250]:
```suggestion
                        checkArgument(delta >= 0 && duration > 0, "stat + "delta is negative");
                        return delta / duration;
```

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 2 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:221]: Can you comment what this class is trying to do?  The class name doesn't seem to describe it well.

File: presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java [hunk 4]
```diff
@@ -243,6 +244,20 @@ public long getPeakNodeTotalMemoryReservationInBytes()
...
+    @JsonProperty
+    @ThriftField(18)
+    public long getTotalCpuTimeInNanos()
+    {
+        return totalCpuTimeInNanos;
+    }
+
+    @JsonProperty
+    @ThriftField(19)
+    public long getTaskAge()
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java:256]: Please add units to all names.

File: presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java [hunk 5]
```diff
@@ -57,22 +57,19 @@
     private final TaskState state;
     private final URI self;
     private final Set<Lifespan> completedDriverGroups;
-
+    private final List<ExecutionFailureInfo> failures;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java:60]: Unrelated change?

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Minor comments

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 6]
```diff
@@ -35,11 +41,20 @@
...
+    private final long cpuStatsWindowSizeInMillis;
 
     @Inject
+    public NodeTaskMap(FinalizerService finalizerService, TaskManagerConfig taskConfig)
+    {
+        this.finalizerService = requireNonNull(finalizerService, "finalizerService is null");
+        this.cpuStatsWindowSizeInMillis = taskConfig.getStatusRefreshMaxWait().toMillis() * 2;
+    }
+
+    @VisibleForTesting
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:53]: this is usually used for package private level access.
Can we make the constructor package private if it is only used for testing.

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 7]
```diff
@@ -35,11 +41,20 @@
...
     public NodeTaskMap(FinalizerService finalizerService)
     {
         this.finalizerService = requireNonNull(finalizerService, "finalizerService is null");
+        this.cpuStatsWindowSizeInMillis = Duration.valueOf("10s").toMillis();
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:57]: just say `10_000` without using `Duration` class?

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 8]
```diff
@@ -49,12 +64,22 @@ public void addTask(InternalNode node, RemoteTask task)
...
     }
 
-    public PartitionedSplitCountTracker createPartitionedSplitCountTracker(InternalNode node, TaskId taskId)
+    public NodeStatsTracker createTaskStatsTrackers(InternalNode node, TaskId taskId)
     {
-        return createOrGetNodeTasks(node).createPartitionedSplitCountTracker(taskId);
+        return createOrGetNodeTasks(node).createTaskStatsTrackers(taskId);
     }
 
     private NodeTasks createOrGetNodeTasks(InternalNode node)
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:85]: can we use something like `nodeTasksMap.computeIfAbsent`

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 9]
```diff
@@ -108,87 +147,158 @@ private void addTask(RemoteTask task)
             }
         }
 
-        public PartitionedSplitCountTracker createPartitionedSplitCountTracker(TaskId taskId)
+        public NodeStatsTracker createTaskStatsTrackers(TaskId taskId)
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:150]: `createTaskStatsTrackers` drop the trailing `s` maybe since we are returning a singular `NodeStatsTracker` .?

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 10]
```diff
@@ -49,12 +64,22 @@ public void addTask(InternalNode node, RemoteTask task)
 
     public int getPartitionedSplitsOnNode(InternalNode node)
     {
-        return createOrGetNodeTasks(node).getPartitionedSplitCount();
+        return (int) createOrGetNodeTasks(node).getPartitionedSplitCount();
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:67]: keep split count as `AtomicInt` maybe to prevent the unsafe cast to `int` ?

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 11]
```diff
@@ -108,87 +147,158 @@ private void addTask(RemoteTask task)
...
+            public synchronized void setValue(long value)
             {
-                if (partitionedSplitCount < 0) {
-                    int oldValue = localPartitionedSplitCount.getAndSet(0);
-                    nodeTotalPartitionedSplitCount.addAndGet(-oldValue);
-                    throw new IllegalArgumentException("partitionedSplitCount is negative");
+                if (value < 0) {
+                    long oldValue = this.value.getAndSet(0L);
+                    totalValue.addAndGet(-oldValue);
+                    throw new IllegalArgumentException(stat + " is negative");
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:188]: Can you help me understand the logic here .
If we are throwing an IllegalArg exception wondering why we need to set a negative value and then throw, shouldn't we just throw immediately.?

File: presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java [hunk 12]
```diff
@@ -243,6 +246,20 @@ public long getPeakNodeTotalMemoryReservationInBytes()
         return peakNodeTotalMemoryReservationInBytes;
     }
 
+    @JsonProperty
+    @ThriftField(18)
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java:250]: ❤️

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 10 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:67]: TaskStatsTracker is using AtomicLong. This would require another class definition.

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 11 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:188]: The value is invalid. We need to fix the accumulated (per node) value for this task first and throw and exception.

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 13]
```diff
@@ -108,87 +129,155 @@ private void addTask(RemoteTask task)
...
         }
+
+        // tracks stats which are passed as accumulated (cpu time) by calculating delta / duration.
+        @ThreadSafe
+        private class AccumulatedTaskStatsTracker
+                extends TaskStatsTracker
+        {
+            private final String stat;
+            private final long windowSizeInMillis;
+            SortedMap<Long, Long> values = new TreeMap<>();
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:210]: Where do we clear this map?  Or is it expected that it continuously grows during the lifetime of the task?

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 14]
```diff
@@ -108,87 +129,155 @@ private void addTask(RemoteTask task)
...
+                super(stat, taskId, totalValue);
+                this.stat = requireNonNull(stat, "stat is null");
+                this.windowSizeInMillis = windowSizeInMillis;
+            }
+
+            private long getDeltaPerSecond(long taskAgeInMillis, long value)
+            {
+                if (value > 0) {
+                    values.put(taskAgeInMillis, value);
+                    values = values.tailMap(values.lastKey() - windowSizeInMillis);
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:223]: The tail map is just a view over the original.  I think this is a memory leak.

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:210]: I wonder if the keys will come into this class in time-order (and hence already sorted).  In which case we could use a LinkedHashMap instead.

File: presto-main/src/main/java/com/facebook/presto/execution/RemoteTaskFactory.java [hunk 15]
```diff
@@ -31,7 +30,7 @@ RemoteTask createRemoteTask(Session session,
             PlanFragment fragment,
             Multimap<PlanNodeId, Split> initialSplits,
             OutputBuffers outputBuffers,
-            PartitionedSplitCountTracker partitionedSplitCountTracker,
+            NodeTaskMap.NodeStatsTracker nodeStatsTracker,
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/RemoteTaskFactory.java:33]: Please static import

File: presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java [hunk 16]
```diff
@@ -237,6 +240,7 @@ public TaskStatus getTaskStatus()
 
     private TaskStatus createTaskStatus(TaskHolder taskHolder)
     {
+        long taskStatusAgeInMilis = MILLISECONDS.convert(System.nanoTime() - creationTime, NANOSECONDS);
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:243]: MILLISECONDS.toNanos

File: presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java [hunk 17]
```diff
@@ -57,23 +57,22 @@
...
-
     private final double outputBufferUtilization;
     private final boolean outputBufferOverutilized;
-
     private final long physicalWrittenDataSizeInBytes;
     private final long memoryReservationInBytes;
     private final long systemMemoryReservationInBytes;
     private final long peakNodeTotalMemoryReservationInBytes;
-
     private final long fullGcCount;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java:68]: Unrelated whitespace change?

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java [hunk 18]
```diff
@@ -518,6 +518,86 @@ public void testSplitCount()
         assertEquals(nodeTaskMap.getPartitionedSplitsOnNode(chosenNode), 0);
     }
 
+    @Test
+    public void testCPUUsage()
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java:522]: Nit: testCpuUsage

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
A few more comments

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:210]: I added the following to avoid memory leak. There is a chance that It can be out of order since updateTaskStats in HttpRemoteTask being called from several places. 

`values.headMap(values.lastKey() - windowSizeInMillis).clear();`

[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:210]: It's still missing the visibility modifier.

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 19]
```diff
@@ -108,87 +130,154 @@ private void addTask(RemoteTask task)
...
+            }
+
+            private long getDeltaPerSecond(long taskAgeInMillis, long value)
+            {
+                if (value > 0) {
+                    values.put(taskAgeInMillis, value);
+                    // this clears the map and make items eligible for GC
+                    values.headMap(values.lastKey() - windowSizeInMillis).clear();
+                    if (values.size() > 1) {
+                        long deltaValue = (values.get(values.lastKey()) - values.get(values.firstKey())) * 100;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:225]: Why multiply by 100?

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java [hunk 20]
```diff
@@ -108,87 +130,154 @@ private void addTask(RemoteTask task)
...
+                        long deltaDuration = values.lastKey() - values.firstKey();
+                        return deltaDuration >= 0 && deltaValue > 0 ? deltaValue / deltaDuration : 0;
+                    }
+                }
+                return 0;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:230]: Can you help me understand why this would return 0 here?

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:210]: Also, do you want to use a fastutil `Long2LongSortedMap` to avoid boxing?

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 19 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:225]: This was there from the initial implementation. It was helping to calculate a percentage. I will remove.

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 20 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:230]: We are return 0 if the reporting is not valid or we dont have enough data points.

File: presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java (hunk 19 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:225]: Actually, using percentage as utilization unit is a common practice. I will keep 100 and change the of the method to get CPU utilization.

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java:225]: There's methods which don't require boxing: e.g. `lastLongKey`, `firstLongKey`

**Review by tdcmeehan — APPROVED** [review]
LGTM

**Comment by mayankgarg1990** [comment]
A late participant to this discussion. 

I am curious if you considered an alternative where the node just sends its system specific stats as compared to aggregating these metrics at the task level. So somehow as a part of the heartbeat from the Node to the coordinator, it sends additional information stating its cpu and memory utilization.

**Comment by tdcmeehan** [comment]
@mayankgarg1990 and @cemcayiroglu (to correct me if I'm wrong), longer term we're planning something like that--not exactly through DiscoveryService heartbeats as the schema is a bit inefficient for this purpose, but rather through the signals we receive in `ClusterMemoryManager` and refactoring that out into a separate service.  Tasks are considered an easy first step to prove out the technique, as it doesn't require significant refactor or introduction of new API, and they are easily aggregated to the node level.

**Comment by cemcayiroglu** [comment]
@tdcmeehan @mayankgarg1990 yes, Tim you are right! This quick change unblocks proving the workload placement technique as a first step. In the long/medium term, we are going to have a separate service that provides this info to `ClusterMemoryManager`, `NodeSelectors`, and potentially UI.
